djautotask/sync_rest.py

```diff
@@ -271,3 +291,130 @@ def _assign_field_data(self, instance, json_data):
+            ]
+
+    def _assign_field_data(self, instance, json_data):
+        instance.id = json_data['id']
+        instance.title = json_data['title']
+
+        instance.ticket_number = json_data.get('ticketNumber')
+        instance.description = json_data.get('description')
+        instance.estimated_hours = json_data.get('estimatedHours')
+        instance.service_level_agreement = \
+            json_data.get('serviceLevelAgreementID')
+        instance.service_level_agreement_has_been_met = \
+            bool(json_data.get('serviceLevelAgreementHasBeenMet'))
+        sla_paused = \
+            json_data.get('serviceLevelAgreementPausedNextEventHours')
+        instance.first_response_date_time = \
+            json_data.get('firstResponseDateTime')
+        instance.first_response_due_date_time = \
+            json_data.get('firstResponseDueDateTime')
+        instance.resolution_plan_date_time = \
+            json_data.get('resolutionPlanDateTime')
+        instance.resolution_plan_due_date_time = \
+            json_data.get('resolutionPlanDueDateTime')
+        instance.resolved_date_time = \
+            json_data.get('resolvedDateTime')
+        instance.resolved_due_date_time = \
+            json_data.get('resolvedDueDateTime')
+        instance.create_date = json_data.get('createDate')
+        instance.due_date_time = json_data.get('dueDateTime')
+        instance.completed_date = json_data.get('completedDate')
+        instance.last_activity_date = json_data.get('lastActivityDate')
+
+        instance = self._parse_date_time(instance, 'first_response_date_time')
+        instance = self._parse_date_time(instance,
+                                         'first_response_due_date_time')
+        instance = self._parse_date_time(instance, 'resolution_plan_date_time')
+        instance = self._parse_date_time(instance,
+                                         'resolution_plan_due_date_time')
+        instance = self._parse_date_time(instance, 'resolved_date_time')
+        instance = self._parse_date_time(instance, 'resolved_due_date_time')
+        instance = self._parse_date_time(instance, 'create_date')
+        instance = self._parse_date_time(instance, 'due_date_time')
+        instance = self._parse_date_time(instance, 'completed_date')
+        instance = self._parse_date_time(instance, 'last_activity_date')
+
+        udfs = json_data.get('userDefinedFields')
+
+        # Refresh udf field to eliminate stale udfs
+        instance.udf = dict()
+
+        if len(udfs):
+            self._assign_udf_data(instance, udfs)
+
+        if sla_paused:
+            instance.service_level_agreement_paused_next_event_hours = \
+                Decimal(str(round(sla_paused, 2)))
+        if instance.estimated_hours:
+            instance.estimated_hours = \
+                Decimal(str(round(instance.estimated_hours, 2)))
+
+        self.set_relations(instance, json_data)
```

Comment:
I'm seeing block of warning logs for almost every ticket.

```
[2021-04-21 14:15:30] INFO [djautotask.sync_rest:195] Skipped: Ticket 9171-Document the Debian                                               
[2021-04-21 14:15:30] WARNING [djautotask.sync_rest:67] Failed to find projectID None for <class 'djautotask.models.TicketTracker'> 9172.    
[2021-04-21 14:15:30] WARNING [djautotask.sync_rest:67] Failed to find source None for <class 'djautotask.models.TicketTracker'> 9172.       
[2021-04-21 14:15:30] WARNING [djautotask.sync_rest:67] Failed to find issueType None for <class 'djautotask.models.TicketTracker'> 9172.    
[2021-04-21 14:15:30] WARNING [djautotask.sync_rest:67] Failed to find subIssueType None for <class 'djautotask.models.TicketTracker'> 9172. 
[2021-04-21 14:15:30] WARNING [djautotask.sync_rest:67] Failed to find billingCodeID None for <class 'djautotask.models.TicketTracker'> 9172.
[2021-04-21 14:15:30] WARNING [djautotask.sync_rest:67] Failed to find contractID None for <class 'djautotask.models.TicketTracker'> 9172.   
[2021-04-21 14:15:30] WARNING [djautotask.sync_rest:67] Failed to find contactID None for <class 'djautotask.models.TicketTracker'> 9172.    
```

I debugged a little bit and it looks like the API returns all fields for a Ticket entity even if the fields aren't set to anything.  In contrast, CW just doesn't return a field in the API if the field is not set/has no value.

The json_data has a lot of keys with value `None`. So I suspect the code in `_assign_relation` is attempting to lookup a lot of `related_instance = model_class.objects.get(pk=uid)` where uid is None.

Reply:
Good point! Thanks Cameron.